File: bindings/lua/src/transaction.lua
local TransactionManager = {}
TransactionManager.__index = TransactionManager

-- Constants
local MAX_INPUTS = 100
local MAX_OUTPUTS = 100
local MIN_FEE = 0.0001
local MAX_TRANSACTION_SIZE = 1024 * 1024  -- 1MB

function TransactionManager.new()
    local self = setmetatable({}, TransactionManager)
    self.initialized = false
    self.blockchainManager = nil
    self.quantumManager = nil
    return self
end

function TransactionManager:initialize(blockchainManager, quantumManager)
    if self.initialized then
        error("TransactionManager already initialized")
    end

    if not blockchainManager then
        error("BlockchainManager is required")
    end

    if not quantumManager then
        error("QuantumManager is required")
    end

    self.blockchainManager = blockchainManager
    self.quantumManager = quantumManager
    self.initialized = true
end

function TransactionManager:createTransaction(inputs, outputs, fee)
    if not self.initialized then
        error("TransactionManager not initialized")
    end

    if not inputs or not outputs then
        error("Inputs and outputs are required")
    end

    if #inputs > MAX_INPUTS then
        error("Too many inputs")
    end

    if #outputs > MAX_OUTPUTS then
        error("Too many outputs")
    end

    fee = fee or MIN_FEE

    -- Calculate total input and output amounts
    local totalInput = 0
    local totalOutput = 0

    for _, input in ipairs(inputs) do
        if not input.amount or input.amount <= 0 then
            error("Invalid input amount")
        end
        totalInput = totalInput + input.amount
    end

    for _, output in ipairs(outputs) do
        if not output.amount or output.amount <= 0 then
            error("Invalid output amount")
        end
        totalOutput = totalOutput + output.amount
    end

    -- Verify sufficient funds
    if totalOutput + fee > totalInput then
        error("Insufficient funds")
    end

    -- Create transaction
    local transaction = {
        hash = self:_generateTransactionHash(),
        inputs = inputs,
        outputs = outputs,
        fee = fee,
        timestamp = os.time(),
        version = 1
    }

    return transaction
end

function TransactionManager:signTransaction(transaction, privateKey)
    if not self.initialized then
        error("TransactionManager not initialized")
    end

    if not transaction or not transaction.hash then
        error("Invalid transaction")
    end

    if not privateKey then
        error("Private key is required")
    end

    -- Create signature data
    local signatureData = self:_createSignatureData(transaction)

    -- Sign the transaction
    local signature = self.quantumManager:encrypt(privateKey, signatureData)
    transaction.signature = signature

    return transaction
end

function TransactionManager:verifyTransaction(transaction)
    if not self.initialized then
        error("TransactionManager not initialized")
    end

    if not transaction or not transaction.hash then
        error("Invalid transaction")
    end

    -- Verify signature
    if not transaction.signature then
        return false
    end

    -- Verify inputs and outputs
    if not self:_verifyInputsOutputs(transaction) then
        return false
    end

    -- Verify transaction size
    if not self:_verifyTransactionSize(transaction) then
        return false
    end

    return true
end

function TransactionManager:submitTransaction(transaction)
    if not self.initialized then
        error("TransactionManager not initialized")
    end

    if not transaction or not transaction.hash then
        error("Invalid transaction")
    end

    -- Verify transaction
    if not self:verifyTransaction(transaction) then
        error("Invalid transaction")
    end

    -- Submit to blockchain
    return self.blockchainManager:addTransaction(transaction)
end

function TransactionManager:getTransactionStatus(hash)
    if not self.initialized then
        error("TransactionManager not initialized")
    end

    if not hash then
        error("Invalid transaction hash")
    end

    -- Check if transaction is in a block
    local transaction = self.blockchainManager:getTransaction(hash)
    if not transaction then
        return "not_found"
    end

    -- Check if transaction is pending
    if self.blockchainManager.pendingTransactions[hash] then
        return "pending"
    end

    return "confirmed"
end

function TransactionManager:_generateTransactionHash()
    -- This is a placeholder for the actual implementation
    -- In a real implementation, this would use a proper hashing algorithm
    return "tx_" .. os.time() .. "_" .. math.random(1000000)
end

function TransactionManager:_createSignatureData(transaction)
    -- Create a string representation of the transaction for signing
    local data = {
        hash = transaction.hash,
        inputs = transaction.inputs,
        outputs = transaction.outputs,
        fee = transaction.fee,
        timestamp = transaction.timestamp,
        version = transaction.version
    }
    return data
end

function TransactionManager:_verifyInputsOutputs(transaction)
    -- Verify input amounts
    local totalInput = 0
    for _, input in ipairs(transaction.inputs) do
        if not input.amount or input.amount <= 0 then
            return false
        end
        totalInput = totalInput + input.amount
    end

    -- Verify output amounts
    local totalOutput = 0
    for _, output in ipairs(transaction.outputs) do
        if not output.amount or output.amount <= 0 then
            return false
        end
        totalOutput = totalOutput + output.amount
    end

    -- Verify total amounts
    if totalOutput + transaction.fee > totalInput then
        return false
    end

    return true
end

function TransactionManager:_verifyTransactionSize(transaction)
    -- This is a placeholder for the actual implementation
    -- In a real implementation, this would calculate the actual size of the transaction
    local size = 0
    -- Add size of all fields
    size = size + #tostring(transaction.hash)
    size = size + #tostring(transaction.fee)
    size = size + #tostring(transaction.timestamp)
    size = size + #tostring(transaction.version)
    size = size + #tostring(transaction.signature)

    for _, input in ipairs(transaction.inputs) do
        size = size + #tostring(input.amount)
        size = size + #tostring(input.address)
    end

    for _, output in ipairs(transaction.outputs) do
        size = size + #tostring(output.amount)
        size = size + #tostring(output.address)
    end

    return size <= MAX_TRANSACTION_SIZE
end

return TransactionManager 

File: src/core/src/transaction.cpp
/**
 * @file $(basename "$1")
 * @brief $(basename "$1" | sed 's/\./_/g' | tr '[:lower:]' '[:upper:]')
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "satox/core/transaction.h"
#include <stdexcept>
#include <sstream>
#include <iomanip>
#include <openssl/sha.h>
#include <openssl/evp.h>
#include <openssl/err.h>
#include <openssl/ec.h>
#include <openssl/bn.h>

namespace satox {
namespace core {

TransactionManager::TransactionManager() : initialized_(false) {}

TransactionManager::~TransactionManager() {
    if (initialized_) {
        cleanup();
    }
}

void TransactionManager::initialize() {
    if (initialized_) {
        throw std::runtime_error("TransactionManager already initialized");
    }
    initialized_ = true;
}

Transaction TransactionManager::createTransaction(
    const std::string& from,
    const std::string& to,
    const std::string& amount,
    const std::string& data) {
    
    if (!initialized_) {
        throw std::runtime_error("TransactionManager not initialized");
    }

    if (from.empty() || to.empty() || amount.empty()) {
        throw std::runtime_error("From, to, and amount are required");
    }

    Transaction transaction;
    transaction.from = from;
    transaction.to = to;
    transaction.amount = amount;
    transaction.data = data;
    transaction.timestamp = std::chrono::system_clock::now();
    transaction.status = "pending";

    // Generate transaction ID
    transaction.id = generateTransactionId(transaction);

    return transaction;
}

bool TransactionManager::signTransaction(Transaction& transaction, const std::string& privateKey) {
    if (!initialized_) {
        throw std::runtime_error("TransactionManager not initialized");
    }

    if (privateKey.empty()) {
        throw std::runtime_error("Private key is required");
    }

    try {
        // Create signature data
        std::stringstream ss;
        ss << transaction.from << transaction.to << transaction.amount 
           << transaction.data << transaction.timestamp.time_since_epoch().count();
        
        std::string dataToSign = ss.str();
        
        // Create signature using ECDSA
        EVP_PKEY* pkey = nullptr;
        EVP_PKEY_CTX* ctx = nullptr;
        
        // Convert private key from hex string to EVP_PKEY
        std::vector<unsigned char> privateKeyBytes;
        if (!hexStringToBytes(privateKey, privateKeyBytes)) {
            throw std::runtime_error("Invalid private key format");
        }
        
        // Create EC key from private key bytes
        EC_KEY* ecKey = EC_KEY_new_by_curve_name(NID_secp256k1);
        if (!ecKey) {
            throw std::runtime_error("Failed to create EC key");
        }
        
        BIGNUM* privBN = BN_bin2bn(privateKeyBytes.data(), privateKeyBytes.size(), nullptr);
        if (!privBN) {
            EC_KEY_free(ecKey);
            throw std::runtime_error("Failed to create BIGNUM from private key");
        }
        
        if (EC_KEY_set_private_key(ecKey, privBN) != 1) {
            BN_free(privBN);
            EC_KEY_free(ecKey);
            throw std::runtime_error("Failed to set private key");
        }
        
        // Generate public key from private key
        if (EC_KEY_generate_key(ecKey) != 1) {
            BN_free(privBN);
            EC_KEY_free(ecKey);
            throw std::runtime_error("Failed to generate public key");
        }
        
        pkey = EVP_PKEY_new();
        if (!pkey) {
            BN_free(privBN);
            EC_KEY_free(ecKey);
            throw std::runtime_error("Failed to create EVP_PKEY");
        }
        
        if (EVP_PKEY_assign_EC_KEY(pkey, ecKey) != 1) {
            EVP_PKEY_free(pkey);
            BN_free(privBN);
            throw std::runtime_error("Failed to assign EC key to EVP_PKEY");
        }
        
        // Create signature context
        ctx = EVP_PKEY_CTX_new(pkey, nullptr);
        if (!ctx) {
            EVP_PKEY_free(pkey);
            BN_free(privBN);
            throw std::runtime_error("Failed to create signature context");
        }
        
        // Initialize signing
        if (EVP_PKEY_sign_init(ctx) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            EVP_PKEY_free(pkey);
            BN_free(privBN);
            throw std::runtime_error("Failed to initialize signing");
        }
        
        // Set digest algorithm
        if (EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha256()) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            EVP_PKEY_free(pkey);
            BN_free(privBN);
            throw std::runtime_error("Failed to set signature digest");
        }
        
        // Calculate signature length
        size_t sigLen;
        if (EVP_PKEY_sign(ctx, nullptr, &sigLen, 
                         reinterpret_cast<const unsigned char*>(dataToSign.c_str()), 
                         dataToSign.length()) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            EVP_PKEY_free(pkey);
            BN_free(privBN);
            throw std::runtime_error("Failed to calculate signature length");
        }
        
        // Create signature
        std::vector<unsigned char> signature(sigLen);
        if (EVP_PKEY_sign(ctx, signature.data(), &sigLen,
                         reinterpret_cast<const unsigned char*>(dataToSign.c_str()),
                         dataToSign.length()) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            EVP_PKEY_free(pkey);
            BN_free(privBN);
            throw std::runtime_error("Failed to create signature");
        }
        
        // Convert signature to hex string
        transaction.signature = bytesToHexString(signature);
        
        // Cleanup
        EVP_PKEY_CTX_free(ctx);
        EVP_PKEY_free(pkey);
        BN_free(privBN);
        
        return true;
    } catch (const std::exception& e) {
        throw std::runtime_error("Failed to sign transaction: " + std::string(e.what()));
    }
}

bool TransactionManager::verifyTransaction(const Transaction& transaction) {
    if (!initialized_) {
        throw std::runtime_error("TransactionManager not initialized");
    }

    if (!validateTransaction(transaction)) {
        return false;
    }

    try {
        // Create signature data (same as in signing)
        std::stringstream ss;
        ss << transaction.from << transaction.to << transaction.amount 
           << transaction.data << transaction.timestamp.time_since_epoch().count();
        
        std::string dataToVerify = ss.str();
        
        // Convert signature from hex string to bytes
        std::vector<unsigned char> signatureBytes;
        if (!hexStringToBytes(transaction.signature, signatureBytes)) {
            return false;
        }
        
        // Derive public key from transaction data
        // In a real implementation, you would get the public key from the transaction
        // For now, we'll use a placeholder approach
        std::string publicKey = derivePublicKeyFromTransaction(transaction);
        
        // Create EVP_PKEY from public key
        EVP_PKEY* pkey = nullptr;
        EVP_PKEY_CTX* ctx = nullptr;
        
        // Convert public key from hex string to EVP_PKEY
        std::vector<unsigned char> publicKeyBytes;
        if (!hexStringToBytes(publicKey, publicKeyBytes)) {
            return false;
        }
        
        // Create EC key from public key bytes
        EC_KEY* ecKey = EC_KEY_new_by_curve_name(NID_secp256k1);
        if (!ecKey) {
            return false;
        }
        
        EC_POINT* pubPoint = EC_POINT_new(EC_KEY_get0_group(ecKey));
        if (!pubPoint) {
            EC_KEY_free(ecKey);
            return false;
        }
        
        if (EC_POINT_oct2point(EC_KEY_get0_group(ecKey), pubPoint,
                              publicKeyBytes.data(), publicKeyBytes.size(), nullptr) != 1) {
            EC_POINT_free(pubPoint);
            EC_KEY_free(ecKey);
            return false;
        }
        
        if (EC_KEY_set_public_key(ecKey, pubPoint) != 1) {
            EC_POINT_free(pubPoint);
            EC_KEY_free(ecKey);
            return false;
        }
        
        pkey = EVP_PKEY_new();
        if (!pkey) {
            EC_POINT_free(pubPoint);
            EC_KEY_free(ecKey);
            return false;
        }
        
        if (EVP_PKEY_assign_EC_KEY(pkey, ecKey) != 1) {
            EVP_PKEY_free(pkey);
            EC_POINT_free(pubPoint);
            EC_KEY_free(ecKey);
            return false;
        }
        
        // Create verification context
        ctx = EVP_PKEY_CTX_new(pkey, nullptr);
        if (!ctx) {
            EVP_PKEY_free(pkey);
            EC_POINT_free(pubPoint);
            return false;
        }
        
        // Initialize verification
        if (EVP_PKEY_verify_init(ctx) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            EVP_PKEY_free(pkey);
            EC_POINT_free(pubPoint);
            return false;
        }
        
        // Set digest algorithm
        if (EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha256()) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            EVP_PKEY_free(pkey);
            EC_POINT_free(pubPoint);
            return false;
        }
        
        // Verify signature
        int result = EVP_PKEY_verify(ctx, signatureBytes.data(), signatureBytes.size(),
                                    reinterpret_cast<const unsigned char*>(dataToVerify.c_str()),
                                    dataToVerify.length());
        
        // Cleanup
        EVP_PKEY_CTX_free(ctx);
        EVP_PKEY_free(pkey);
        EC_POINT_free(pubPoint);
        
        return result == 1;
    } catch (const std::exception& e) {
        return false;
    }
}

bool TransactionManager::broadcastTransaction(const Transaction& transaction) {
    if (!initialized_) {
        throw std::runtime_error("TransactionManager not initialized");
    }

    if (!verifyTransaction(transaction)) {
        return false;
    }

    // Add to pending transactions
    pendingTransactions_.push_back(transaction);

    // Broadcast to network
    // This is a placeholder implementation
    // Replace with actual broadcasting logic
    return true;
}

std::string TransactionManager::getTransactionStatus(const std::string& transactionId) {
    if (!initialized_) {
        throw std::runtime_error("TransactionManager not initialized");
    }

    if (transactionId.empty()) {
        throw std::runtime_error("Transaction ID is required");
    }

    // Check pending transactions
    for (const auto& tx : pendingTransactions_) {
        if (tx.id == transactionId) {
            return tx.status;
        }
    }

    // Check transaction history
    for (const auto& tx : transactionHistory_) {
        if (tx.id == transactionId) {
            return tx.status;
        }
    }

    return "not_found";
}

std::vector<Transaction> TransactionManager::getPendingTransactions() {
    if (!initialized_) {
        throw std::runtime_error("TransactionManager not initialized");
    }

    return pendingTransactions_;
}

std::vector<Transaction> TransactionManager::getTransactionHistory(const std::string& address) {
    if (!initialized_) {
        throw std::runtime_error("TransactionManager not initialized");
    }

    if (address.empty()) {
        throw std::runtime_error("Address is required");
    }

    std::vector<Transaction> history;
    for (const auto& tx : transactionHistory_) {
        if (tx.from == address || tx.to == address) {
            history.push_back(tx);
        }
    }

    return history;
}

void TransactionManager::cleanup() {
    if (!initialized_) {
        return;
    }

    pendingTransactions_.clear();
    transactionHistory_.clear();
    initialized_ = false;
}

std::string TransactionManager::generateTransactionId(const Transaction& transaction) {
    std::stringstream ss;
    ss << transaction.from << transaction.to << transaction.amount 
       << transaction.data << transaction.timestamp.time_since_epoch().count();

    // Generate SHA-256 hash
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, ss.str().c_str(), ss.str().size());
    SHA256_Final(hash, &sha256);

    // Convert to hex string
    std::stringstream hexStream;
    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        hexStream << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(hash[i]);
    }

    return hexStream.str();
}

bool TransactionManager::validateTransaction(const Transaction& transaction) {
    if (transaction.from.empty() || transaction.to.empty() || 
        transaction.amount.empty() || transaction.id.empty()) {
        return false;
    }

    // Add more validation as needed
    return true;
}

// Helper methods
bool TransactionManager::hexStringToBytes(const std::string& hex, std::vector<unsigned char>& bytes) {
    if (hex.length() % 2 != 0) {
        return false;
    }
    
    bytes.clear();
    bytes.reserve(hex.length() / 2);
    
    for (size_t i = 0; i < hex.length(); i += 2) {
        std::string byteString = hex.substr(i, 2);
        try {
            unsigned char byte = static_cast<unsigned char>(std::stoi(byteString, nullptr, 16));
            bytes.push_back(byte);
        } catch (const std::exception&) {
            return false;
        }
    }
    
    return true;
}

std::string TransactionManager::bytesToHexString(const std::vector<unsigned char>& bytes) {
    std::stringstream ss;
    ss << std::hex << std::setfill('0');
    for (unsigned char byte : bytes) {
        ss << std::setw(2) << static_cast<int>(byte);
    }
    return ss.str();
}

std::string TransactionManager::derivePublicKeyFromTransaction(const Transaction& transaction) {
    // In a real implementation, this would derive the public key from the transaction data
    // For now, we'll return a placeholder public key
    // This should be replaced with actual public key derivation logic
    
    // Create a deterministic public key from transaction data
    std::stringstream ss;
    ss << transaction.from << transaction.to << transaction.amount;
    std::string data = ss.str();
    
    // Hash the data to create a deterministic public key
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, data.c_str(), data.length());
    SHA256_Final(hash, &sha256);
    
    // Convert to hex string (this is a simplified approach)
    std::stringstream hexStream;
    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        hexStream << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(hash[i]);
    }
    
    return hexStream.str();
}

} // namespace core
} // namespace satox 